### ISSUE TITLE
perf: A* DeepCubeA 1.438s → ~2.20s from JAX 0.8.1→0.8.2 + Xtructure a37641b

### ISSUE THREAD

**Issue by tinker495** [tinker495/JAxtar#8]
# [perf/deps] A* DeepCubeA 1.438s → ~2.20s 회귀 — JAX 0.8.1→0.8.2 패치 + Xtructure `a37641b` 의 누적 효과

## 요약 (TL;DR)

`docs/benchmarks/rubikscube-deepcubea.md:28` 의 docs row:
```
A* | Diffusion Distance - 14.7M params | 100% / 54.60% | 1.438s | 1.82M | 21.61 | 0.891 | 0.952
```

같은 명령이 현재 같은 머신/같은 워크로드에서 `~2.20s` 로 측정됩니다. 31 회 wall-clock 측정으로 분해해 본 결과, **JAxtar 자체의 어떤 commit 도 회귀 원인이 아닙니다**. 회귀는 두 의존성 변화의 누적 효과입니다:

1. **Xtructure `a37641b` (2026-01-25, "single fingerprint hashtable lookup + insert_triton.py 도입")** — JAX 0.8.x 위에서 **+0.384s** Xtructure-side 회귀. 별도로 JAX ≤ 0.7.x 호환성을 깨뜨림 (`pl_triton.atomic_or` 의존). [Xtructure issue 별도]
2. **JAX 0.8.1 → 0.8.2 minor 패치** — HEAD Xtructure 코드 위에서 **+0.512s** JAX-side 회귀 (메인 mission 측정).

두 회귀는 부분적으로 overlap (JAX 0.9.x 가 둘을 가리는 효과) 으로 실측 합 ≈ +0.76s.

## 측정 데이터 — JAX 축 (JAxtar HEAD + xt HEAD + PuXle HEAD, 모두 main)

| jax | mean wall-clock (s) | 클러스터 |
| :--- | ---: | :---: |
| 0.8.0 | 1.580 | fast |
| 0.8.1 | 1.587 | fast |
| **0.8.2** | **2.099** | **slow ← 메인 jump (+0.512)** |
| 0.8.3 | 2.071 | slow |
| 0.9.0.1 | 2.197 | slow |
| 0.9.2 | 2.137 | slow |
| 0.7.0 | (incompatible) | xt HEAD 의 `pl_triton.atomic_or` 미존재 |

**bracket 검증:** [4, 5] 인접, gap 0.512s ≥ 0.5s 임계, 양쪽 클러스터 내 spread ≤ 0.03s. autoresearch evaluator pass.

## 측정 데이터 — Xtructure 축 (jax 0.8.1, JAxtar=`4fb2a800`, PuXle=`8f50034` 고정)

| xt SHA | 날짜 | mean (s) | 클러스터 |
| :--- | :--- | ---: | :---: |
| `eee7a6b` | 2025-12-19 | 1.484 | fast |
| `6fbdf2e` | 2026-01-11 | 1.427 | fast |
| `080b291` | 2026-01-11 | 1.396 | fast |
| `b1e9a4f` | 2026-01-12 | 1.457 | fast |
| `13aeebb` | 2026-01-14 | 1.428 | fast |
| `2b39104` | 2026-01-24 | 1.370 | fast |
| **`a37641b`** | **2026-01-25** | **1.754** | **slow ← xt jump (+0.384)** |
| `e8d332a` | 2026-01-25 | 1.868 | slow |
| `5508b91` | 2026-02-22 | 1.789 | slow |
| `1a17fd1` | 2026-03-02 | 1.744 | slow |
| `45723c1` (HEAD) | 2026-03-03 | 1.775 | slow |

## 측정 데이터 — docs row 재현

다음 조합들이 docs row `1.438s` 를 ±0.10s 안에 재현:

| 조합 | mean (s) |
| :--- | ---: |
| jax 0.7.0 + xt `b1e9a4f` + JAxtar `4fb2a800` + PuXle `8f50034` | 1.298 ⭐ |
| jax 0.7.0 + xt `eee7a6b` + JAxtar `4fb2a800` + PuXle `8f50034` | 1.413 |
| jax 0.7.1 + xt `eee7a6b` + 동일 | 1.418 |
| jax 0.8.1 + xt `eee7a6b` + 동일 | 1.367 (extension_07x) |
| jax 0.8.1 + xt `2b39104` + 동일 | 1.370 |

## JAxtar 자체 책임 — 없음

JAxtar `4fb2a800` (docs-era) ↔ JAxtar HEAD (`0e9190be`) 차이 측정:

- jax 0.8.1 + xt 45723c1: 1.587s vs 1.775s ≈ -0.19s (JAxtar HEAD 가 더 빠름)
- jax 0.9.0.1 + xt 45723c1: 2.197s vs 2.158s ≈ -0.04s (사실상 동일)

JAxtar HEAD 은 4fb2a800 보다 약간 빠르거나 동일. JAxtar 코드는 회귀 원인이 아닙니다.

## 환경

- Python 3.12.12, NumPy 2.4.2, Flax 0.12.2, Optax 0.2.8.dev0
- GPU: NVIDIA RTX 4080 SUPER, driver 591.86
- CUDA 12.x
- `XLA_PYTHON_CLIENT_PREALLOCATE=false`

### Priority 1 — `requirements_gpu.txt` 의 `jax` 핀 추가

현재:
```
jax[cuda]
```

권장:
```
jax[cuda]>=0.8.0,<0.8.2
```
또는 (더 보수적):
```
jax[cuda]==0.8.1
```

근거: jax 0.8.1 위에서 ~1.59s, 0.8.2+ 위에서 ~2.10s. 0.5s 차이의 즉시 회복.

대안: 핀 대신 README 에 "성능 최적 영역: jax==0.8.1, Xtructure ≤ 2b39104" 명기 + `EvalOptions` 의 `max_node_size` 동적 결정으로 capacity 비용을 줄이는 기존 권고도 함께.

### Priority 2 — Xtructure 의존성 핀 또는 fallback 사용 명기

Xtructure HEAD 의 `insert_triton.py` 경로가 jax 0.8.x 위에서 +0.38s 비용을 만듭니다 (Xtructure issue 참조). 두 가지 선택:

1. Xtructure 를 `2b39104` 이전으로 핀 (모듈 분리 이전 코드, 새 기능 사용처 영향 검토 필요)
2. Xtructure 가 `XTRUCTURE_HASHTABLE_INSERT_TRITON=0` opt-in 으로 변경되면 (Xtructure 측 PR 가능) 이 옵션을 default 로 사용

### Priority 3 — docs row 작성 환경 명기

`docs/benchmarks/rubikscube-deepcubea.md` 가 측정된 정확한 환경을 row 옆에 기록하는 컬럼 추가:

| Algorithm | Model | ... | jax | jaxlib | xt sha | Date |

이렇게 하면 미래 회귀 발생 시 즉시 비교 가능.

### Priority 4 — `eval_runner` 에 환경 자동 캡쳐

각 bench run 의 `config.json` 에 `jax.__version__`, `jaxlib.__version__`, `xtructure.__version__`, `puxle.__version__`, `git rev-parse` (가능시) 기록. 이미 일부 메타데이터는 있지만 의존성 버전은 누락.

## 검증 재현 (1 sample)

```bash
# A. 현재 (slow) 측정
conda activate py312  # or env with jax 0.9.0.1 + xt HEAD
cd /path/to/JAxtar
XLA_PYTHON_CLIENT_PREALLOCATE=false python main.py benchmark astar \
  --benchmark rubikscube-deepcubea --sample-ids 0,1,2,3,4,5,6,7,8,9 \
  --batch-size 10000 --max-node-size 20000000 --cost-weight 0.6 \
  --pop_ratio inf --use-early-stopping False
# Expected: ~2.20s

# B. JAX 0.8.1 핀만 적용 (Xtructure HEAD 유지)
pip install --force-reinstall jax[cuda12]==0.8.1
# 같은 명령 → ~1.59s

# C. Xtructure 도 docs-era 로 (전체 회복)
git checkout eee7a6b in xtructure dir, pip install -e .
# 같은 명령 → ~1.37s (docs row 재현)
```

## 관련 자료

- 통합 보고서: `JAX_XTRUCTURE_REGRESSION_REPORT.md` (저장소 root 에 생성됨)
- Raw 데이터: `.omc/autoresearch/jax-version-axis-validation/runs/20260504-005613/`
- Xtructure 측 issue: 별도 첨부 (`issue-xtructure.md`)
- Prior 사이클 verdict (코드 noise 결론): `.omx/specs/autoresearch-jaxtar-perf-cause-experiments/branch-commit-bisect-report.md` — **본 사이클 결과로 갱신됨** (코드 = noise 는 jax 0.9.x 에서만 사실)

## 라벨 제안

`bug`, `performance`, `dependencies`, `documentation`
